### ISSUE TITLE
invoking msbuild -property:TargetFramework=XXX causes problems if only a single target framework is defined in csproj files

### ISSUE THREAD

**Issue by jspelletier** [dotnet/msbuild#8142]
### Issue Description

When invoking msbuild with -property:TargetFramework=XXX (for example net5.0) and the projects in the solution have only
a single target framework defined(for example <TargetFramework>net5.0</TargetFramework>), then this will trigger multiple concurrent builds and can cause file access problems

### Steps to Reproduce

I've included a simplified projects setup to reproduce the problem.
See the readme.md in the zip for details about msbuild commandline I used to reproduce the problem. The important part is -property:TargetFramework=.net5.0

[repro.zip](https://github.com/dotnet/msbuild/files/9955182/repro.zip)

### Expected Behavior
msbuild shouldn't build twice the requested configuration even if only a single target framework is defined.

### Actual Behavior

E:\github\github-sharpmake>"C:\Program Files\Microsoft Visual Studio\2022\Professional\Msbuild\Current\Bin\amd64\MSBuild.exe" -nologo -property:TargetFramework=net5.0 -property:Configuration=debug  -maxcpucount -verbosity:m E:\repro\consoleapp1\ConsoleApp1\ConsoleApp1.sln /t:Rebuild
  ClassLibrary2 -> E:\repro\consoleapp1\ClassLibrary2\bin\Debug\net5.0\ClassLibrary2.dll
  ClassLibrary2 -> E:\repro\consoleapp1\ClassLibrary2\bin\Debug\net5.0\ClassLibrary2.dll
  ClassLibrary1 -> E:\repro\consoleapp1\ClassLibrary1\bin\Debug\net5.0\ClassLibrary1.dll
  ClassLibrary1 -> E:\repro\consoleapp1\ClassLibrary1\bin\Debug\net5.0\ClassLibrary1.dll
  ConsoleApp1 -> E:\repro\consoleapp1\ConsoleApp1\bin\Debug\net5.0\ConsoleApp1.dll

ClassLibrary1 and 2 are built twice.

### Analysis
No idea

### Versions & Configurations
I think the problem can occur on any platform. I've seen build errors caused by this on macos and windows.

reproduced locally using pretty much latest msbuild. 

Microsoft (R) Build Engine version 17.2.1+52cd2da31 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

17.2.1.25201

Problem occured on macos when building using dotnet build on github ci(not sure which version)

**Comment by BenVillalobos** [comment]
Investigation Notes:

This is a conflict between the solution build and the projectreference protocol.

The solution build is building each project with global property `TargetFramework=foo`, while the projectreference protocol (via ConsoleApp1 -> ClassLib -> ..) specifically undefines `TargetFramework` during the build. This leads to a new evaluation, project instance, and new build.

As for a solution, the ideal behavior is clearly that each project only gets built once. The only way to do that would either be to change how TargetFramework is passed during solution builds or changing the P2P protocol.

As for a workaround, if you build your projects starting with `ConsoleApp1.csproj`, you don't get any overbuild from the sln. @forgind experimented with adding `RemoveProperties=TargetFramework` to the MSBuild calls in the `.sln.metaproj`, and that worked around this issue but is by no means ideal.


Side note: `GetReferenceNearestTargetFrameworkTask` could use more logging. I can see that it's undefining targetframework on projects, but not _why_.

**Comment by jspelletier** [comment]
Hi,

But why is there no problems if the projects contains two frameworks versions and don't build any extra projects if I specify a target framework argument on the commandline?

Sadly, I don't think I can use this workaround as the sln contains several projects and some without direct relations to others(plugins) so I must build using the sln.

However, I can remove the target framework argument when we have only a single framework defined in one of the csproj files of the sln but while this workaround works, someone could easily add back the argument in ci scripts without knowing about this and cause random build errors.

Thanks for the quick investigation. Was expecting weeks before someone checks this.

**Comment by BenVillalobos** [comment]
> But why is there no problems if the projects contains two frameworks versions and don't build any extra projects if I specify a target framework argument on the commandline?

This is likely due to the project ref protocol again, but I'm not 100% sure on the details here. I'm assuming you mean your projects have `<TargetFrameworks>foo;bar</>`, and you build via `msbuild /p:targetframework=foo`? Does every project have targetframeworks instead of targetframework?

I assume this is because of the whole "inner vs outer build" logic we have when a project specifies `TargetFrameworks` over `TargetFramework`

**Comment by Forgind** [comment]
If building a project without the global property TargetFramework set would yield a different result than building with it set (as in this case, since building without it would mean foo;bar, whereas with it, it's just foo) then we have to keep the global property 100% of the time—whether building with the .sln or via ProjectReference. That means it tries to build the same thing twice and exits early.

In other words, if you specify multiple TargetFrameworks, then we need to keep the global property.
If you specify no TargetFramework, then we need to keep the global property.
Only if you specify one TargetFramework, and it matches the global property can we dispense with it.

You happen to be hitting case 3, which is unfortunate. If you add other TargetFrameworks or remove it, I think you'll be ok.

**Comment by jspelletier** [comment]
@BenVillalobos 
The problem that I reported is when the csproj files in the sln have only a single framework version specified. For example,<TargetFramework>net5.0</TargetFramework> and you specify that framework version on the command line. However, if the csproj files have the plural version(TargetFrameworks) there is no problems if the argument is specified to restrict the build to a single framework version.

My guess is since this works when <TargetFrameworks>net5.0;net6.0</TargetFrameworks> is used this should be quite easy to fix.

**Comment by Forgind** [comment]
The workaround is very straightforward, but the best fix I know of is more complicated and would hurt performance a little.

**Comment by BenVillalobos** [comment]
I did some digging, turns out the P2P protocol _always_ undefines `TargetFramework ` when a project defines it. The goal there is to avoid an extra evaluation.

https://github.com/dotnet/msbuild/blob/7a8b3369d4c368c01e9790fe26923e10829ca3a4/src/Tasks/Microsoft.Common.CurrentVersion.targets#L1881-L1887

So the core problem is solution builds have no concept of the P2P protocol and naively pass whatever global props you tell it to. A possible solution for this would be to have CLI solution builds have an equivalent `GetTargetFrameworks` call to extract information and define/undefine specific global props.

Given that this behavior is expected, you've got three options:
- Don't pass TF via command line, and let the build do its thing. Not ideal if you want a specific TF to be built.
- Pass TF via command line, but make sure each project has `TargetFrameworks` defined (even if it just has one TF). Projects that aren't built by the solution are safe to have `TargetFramework` defined.
- Create a separate entrypoint project (a dirs.proj) that you use for CLI builds. This project would have to manually `ProjectReference` each project that would be built, but gets around the naive sln build.

**Comment by Forgind** [comment]
> Don't pass TF via command line, and let the build do its thing. Not ideal if you want a specific TF to be built.

What downside do you see to this? If you specify a TF in the project file, we'll build it, just the same as if you specify it on the command line. If those two disagree, you're doing something wrong, and it will likely error.

**Comment by BenVillalobos** [comment]
@Forgind It's not ideal if we're talking about projects with TFs `net5.0;net6.0` defined and you only want one to be built.

**Comment by Forgind** [comment]
If you have multiple target frameworks, and you specify it on the command line, you don't see this bug. So I see this as:
If you only have one TF, never specify it on the command line.
If you have multiple, you can specify it or not as you wish.

Both of those should be fully functional setups.

**Comment by BenVillalobos** [comment]
Right, but the second case doesn't play well with solutions so I renamed the issue to make it more specific.

**Comment by jspelletier** [comment]
I don't know how this is feasible in this codebase but If the non plural TargetFramework is specified in csproj, and a target framework is specified on commandline msbuild should validate the coherence(must be identical to what's in csproj) and display an error if param is different from what is in csproj files. Otherwise, the commandline param should be ignored.
